### ISSUE TITLE
ArrayIndexOutOfBoundsException when creating distinct keys on RCFile

### ISSUE THREAD

**Issue by jinyangli34** [trinodb/trino#18416]
Hitting this exception with `select * from rc_table` On Version 415:

```
io.trino.spi.TrinoException: Failed to read RC file: s3.....
	at io.trino.plugin.hive.rcfile.RcFilePageSource$RcFileBlockLoader.load(RcFilePageSource.java:218)
	at io.trino.spi.block.LazyBlock$LazyData.load(LazyBlock.java:406)
	at io.trino.spi.block.LazyBlock$LazyData.getFullyLoadedBlock(LazyBlock.java:385)
	at io.trino.spi.block.LazyBlock.getLoadedBlock(LazyBlock.java:292)
	at io.trino.spi.Page.getLoadedPage(Page.java:233)
	at io.trino.operator.TableScanOperator.getOutput(TableScanOperator.java:304)
	at io.trino.operator.Driver.processInternal(Driver.java:402)
	at io.trino.operator.Driver.lambda$process$8(Driver.java:305)
	at io.trino.operator.Driver.tryWithLock(Driver.java:701)
	at io.trino.operator.Driver.process(Driver.java:297)
	at io.trino.operator.Driver.processForDuration(Driver.java:268)
	at io.trino.execution.SqlTaskExecution$DriverSplitRunner.processFor(SqlTaskExecution.java:888)
	at io.trino.execution.executor.PrioritizedSplitRunner.process(PrioritizedSplitRunner.java:187)
	at io.trino.execution.executor.TaskExecutor$TaskRunner.run(TaskExecutor.java:561)
	at io.trino.$gen.Trino_415_0____20230725_213857_2.run(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: java.lang.ArrayIndexOutOfBoundsException: Index 170 out of bounds for length 128
	at io.trino.hive.formats.DistinctMapKeys.selectDistinctKeys(DistinctMapKeys.java:71)
	at io.trino.hive.formats.encodings.text.MapEncoding.decodeValueInto(MapEncoding.java:94)
	at io.trino.hive.formats.encodings.text.BlockEncoding.decodeColumn(BlockEncoding.java:66)
	at io.trino.hive.formats.rcfile.RcFileReader$Column.readBlock(RcFileReader.java:585)
	at io.trino.hive.formats.rcfile.RcFileReader.readBlock(RcFileReader.java:438)
	at io.trino.plugin.hive.rcfile.RcFilePageSource$RcFileBlockLoader.load(RcFilePageSource.java:212)
	... 17 more
```

```
47    public boolean[] selectDistinctKeys(Block keyBlock)
48           throws DuplicateMapKeyException
49    {
50        int keyCount = keyBlock.getPositionCount();                       // 91
51        int hashTableSize = keyCount * HASH_MULTIPLIER;          // 182
52
53        if (distinctBuffer.length < keyCount) {
54            int bufferSize = calculateBufferSize(keyCount);              // 128
55            distinctBuffer = new boolean[bufferSize];
56            hashTableBuffer = new int[bufferSize];
57        }
58        boolean[] distinct = distinctBuffer;
59        Arrays.fill(distinct, false);
60        int[] hashTable = hashTableBuffer;
61        Arrays.fill(hashTable, -1);
62
63        int hashTableOffset = 0;
64        for (int i = 0; i < keyCount; i++) {
65            // Nulls are not marked as distinct and thus are ignored
66            if (keyBlock.isNull(i)) {
67                continue;
68            }
69            int hash = getHashPosition(keyBlock, i, hashTableSize);        // 170
70            while (true) {
71                if (hashTable[hashTableOffset + hash] == -1) {                ==> exception: 170 + 0 > 128
72                    hashTable[hashTableOffset + hash] = i;
73                    distinct[i] = true;
74                    break;
75                }
```

Feel like `HASH_MULTIPLIER` is missed for `calculateBufferSize` in line 54?

**Comment by raunaqmorarka** [comment]
cc: @dain

**Comment by dain** [comment]
I think this was already fixed: #19151
